be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/MessageDecoder.scala

```diff
@@ -2,15 +2,31 @@ package ch.epfl.pop.pubsub.graph
 
 import akka.NotUsed
 import akka.stream.scaladsl.Flow
-import ch.epfl.pop.model.network.{JsonRpcRequest, MethodType}
+import ch.epfl.pop.model.network.{JsonRpcRequest, JsonRpcResponse}
+import ch.epfl.pop.jsonNew.HighLevelProtocol._
+import spray.json._
+
+import scala.util.{Success, Try}
 
 object MessageDecoder {
 
-  // Message or TextMessage
   val jsonRpcParser: Flow[Either[JsonString, PipelineError], GraphMessage, NotUsed] = Flow[Either[JsonString, PipelineError]].map {
-    case Left(jsonString) =>
-      Left(JsonRpcRequest("2.0", MethodType.INVALID, ???, Some(0)))
-    // case _ => _ // TODO check if type ok. Else prob Right(pipelineError)
+    case Left(jsonString) => Try(jsonString.parseJson.asJsObject) match {
+      case Success(obj) =>
+        val fields: Set[String] = obj.fields.keySet
+
+        if (fields.contains("method")) { // FIXME check that the error (if any) is correctly propagated
```

Comment:
I don't understand the FIXME here

Reply:
This is actually for the if-else below. AFAIK the `.convertTo` could fail ==> create an error which would be piped into Lefft(_)

```scala
        if (fields.contains("method")) { // FIXME check that the error (if any) is correctly propagated
          Left(obj.convertTo[JsonRpcRequest])
        } else {
          Left(obj.convertTo[JsonRpcResponse])
        }
```